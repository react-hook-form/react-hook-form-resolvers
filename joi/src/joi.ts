import { appendErrors } from 'react-hook-form';
import { toNestObject } from '@hookform/resolvers';
import * as Joi from 'joi';
import { convertArrayToPathName } from '@hookform/resolvers';
import { Resolver } from './types';

const parseErrorSchema = (
  error: Joi.ValidationError,
  validateAllFieldCriteria: boolean,
) =>
  Array.isArray(error.details)
    ? error.details.reduce(
        (previous: Record<string, any>, { path, message = '', type }) => {
          const currentPath = convertArrayToPathName(path);

          return {
            ...previous,
            ...(path
              ? previous[currentPath] && validateAllFieldCriteria
                ? {
                    [currentPath]: appendErrors(
                      currentPath,
                      validateAllFieldCriteria,
                      previous,
                      type,
                      message,
                    ),
                  }
                : {
                    [currentPath]: previous[currentPath] || {
                      message,
                      type,
                      ...(validateAllFieldCriteria
                        ? {
                            types: { [type]: message || true },
                          }
                        : {}),
                    },
                  }
              : {}),
          };
        },
        {},
      )
    : [];

export const joiResolver: Resolver = (
  schema,
  schemaOptions = {
    abortEarly: false,
  },
  { mode } = { mode: 'async' },
) => async (values, context, { criteriaMode, fields }) => {
  try {
    let result;
    if (mode === 'async') {
      result = await schema.validateAsync(values, {
        ...schemaOptions,
        context,
      });
    } else {
      const { value, error } = schema.validate(values, {
        ...schemaOptions,
        context,
      });

      if (error) {
        throw error;
      }

      result = value;
    }

    return {
      values: result,
      errors: {},
    };
  } catch (e) {
    return {
      values: {},
      errors: toNestObject(parseErrorSchema(e, criteriaMode === 'all'), fields),
    };
  }
};
