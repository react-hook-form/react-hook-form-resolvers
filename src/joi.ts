import { appendErrors, transformToNestObject } from 'react-hook-form';
import convertArrayToPathName from './utils/convertArrayToPathName';

type JoiError = {
  details: any;
};

type FieldValues = Record<string, any>;

type FieldError = {
  path: (string | number)[];
  message: string;
  type: string;
};

const parseErrorSchema = (
  error: JoiError,
  validateAllFieldCriteria: boolean,
) => {
  console.log(error.details[0].path);
  return Array.isArray(error.details)
    ? error.details.reduce(
        (previous: FieldValues, { path, message = '', type }: FieldError) => {
          const currentPath = convertArrayToPathName(path);

          return {
            ...previous,
            ...(currentPath
              ? previous[currentPath] && validateAllFieldCriteria
                ? {
                    [currentPath]: appendErrors(
                      currentPath,
                      validateAllFieldCriteria,
                      previous,
                      type,
                      message,
                    ),
                  }
                : {
                    [currentPath]: previous[currentPath] || {
                      message,
                      type,
                      ...(validateAllFieldCriteria
                        ? {
                            types: { [type]: message || true },
                          }
                        : {}),
                    },
                  }
              : {}),
          };
        },
        {},
      )
    : [];
};

export const joiResolver = (
  validationSchema: any,
  config: any = {
    abortEarly: false,
  },
) => async (data: any, _: any = {}, validateAllFieldCriteria = false) => {
  const { error: errors, value: values } = validationSchema.validate(data, {
    ...config,
  });

  return {
    values,
    errors: parseErrorSchema(errors, validateAllFieldCriteria),
  };

  try {
    return {
      values: validationSchema.validate(data, {
        ...config,
      }),
      errors: {},
    };
  } catch (e) {
    return {
      values: {},
      errors: transformToNestObject(
        parseErrorSchema(e, validateAllFieldCriteria),
      ),
    };
  }
};
