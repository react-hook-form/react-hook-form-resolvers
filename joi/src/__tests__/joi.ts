import * as Joi from 'joi';
import { joiResolver } from '..';

const schema = Joi.object({
  username: Joi.string().alphanum().min(3).max(30).required(),
  password: Joi.string().pattern(new RegExp('^[a-zA-Z0-9]{3,30}$')).required(),
  repeatPassword: Joi.ref('password'),
  accessToken: [Joi.string(), Joi.number()],
  birthYear: Joi.number().integer().min(1900).max(2013),
  email: Joi.string().email({
    minDomainSegments: 2,
    tlds: { allow: ['com', 'net'] },
  }),
  tags: Joi.array().items(Joi.string()).required(),
  enabled: Joi.boolean().required(),
});

interface Data {
  username: string;
  password: string;
  repeatPassword: string;
  accessToken?: number | string;
  birthYear?: number;
  email?: string;
  tags: string[];
  enabled: boolean;
}

describe('joiResolver', () => {
  it('should return values from joiResolver when validation pass', async () => {
    const data: Data = {
      username: 'Doe',
      password: 'Password123',
      repeatPassword: 'Password123',
      birthYear: 2000,
      email: 'john@doe.com',
      tags: ['tag1', 'tag2'],
      enabled: true,
    };

    const validateAsyncSpy = jest.spyOn(schema, 'validateAsync');
    const validateSpy = jest.spyOn(schema, 'validate');

    const result = await joiResolver(schema)(data);

    expect(validateSpy).not.toHaveBeenCalled();
    expect(validateAsyncSpy).toHaveBeenCalledTimes(1);
    expect(result).toEqual({ errors: {}, values: data });
  });

  it('should return values from joiResolver with `mode: sync` when validation pass', async () => {
    const data: Data = {
      username: 'Doe',
      password: 'Password123',
      repeatPassword: 'Password123',
      birthYear: 2000,
      email: 'john@doe.com',
      tags: ['tag1', 'tag2'],
      enabled: true,
    };

    const validateAsyncSpy = jest.spyOn(schema, 'validateAsync');
    const validateSpy = jest.spyOn(schema, 'validate');

    const result = await joiResolver(schema, undefined, { mode: 'sync' })(data);

    expect(validateAsyncSpy).not.toHaveBeenCalled();
    expect(validateSpy).toHaveBeenCalledTimes(1);
    expect(result).toEqual({ errors: {}, values: data });
  });

  it('should return a single error from joiResolver when validation fails', async () => {
    const data = {
      password: '___',
      email: '',
      birthYear: 'birthYear',
    };

    const result = await joiResolver(schema)(data);

    expect(result).toMatchSnapshot();
  });

  it('should return a single error from joiResolver with `mode: sync` when validation fails', async () => {
    const data = {
      password: '___',
      email: '',
      birthYear: 'birthYear',
    };

    const validateAsyncSpy = jest.spyOn(schema, 'validateAsync');
    const validateSpy = jest.spyOn(schema, 'validate');

    const result = await joiResolver(schema, undefined, { mode: 'sync' })(data);

    expect(validateAsyncSpy).not.toHaveBeenCalled();
    expect(validateSpy).toHaveBeenCalledTimes(1);
    expect(result).toMatchSnapshot();
  });

  it('should return all the errors from joiResolver when validation fails with `validateAllFieldCriteria` set to true', async () => {
    const data = {
      password: '___',
      email: '',
      birthYear: 'birthYear',
    };

    const result = await joiResolver(schema)(data, undefined, true);

    expect(result).toMatchSnapshot();
  });

  it('should return all the errors from joiResolver when validation fails with `validateAllFieldCriteria` set to true and `mode: sync`', async () => {
    const data = {
      password: '___',
      email: '',
      birthYear: 'birthYear',
    };

    const result = await joiResolver(schema, undefined, { mode: 'sync' })(
      data,
      undefined,
      true,
    );

    expect(result).toMatchSnapshot();
  });
});
