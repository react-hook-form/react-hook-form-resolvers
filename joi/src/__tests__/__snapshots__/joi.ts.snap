// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`joiResolver should return a single error from joiResolver when validation fails 1`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "\\"birthYear\\" must be a number",
      "type": "number.base",
    },
    "email": Object {
      "message": "\\"email\\" is not allowed to be empty",
      "type": "string.empty",
    },
    "enabled": Object {
      "message": "\\"enabled\\" is required",
      "type": "any.required",
    },
    "password": Object {
      "message": "\\"password\\" with value \\"___\\" fails to match the required pattern: /^[a-zA-Z0-9]{3,30}$/",
      "type": "string.pattern.base",
    },
    "tags": Object {
      "message": "\\"tags\\" is required",
      "type": "any.required",
    },
    "username": Object {
      "message": "\\"username\\" is required",
      "type": "any.required",
    },
  },
  "values": Object {},
}
`;

exports[`joiResolver should return a single error from joiResolver with \`mode: sync\` when validation fails 1`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "\\"birthYear\\" must be a number",
      "type": "number.base",
    },
    "email": Object {
      "message": "\\"email\\" is not allowed to be empty",
      "type": "string.empty",
    },
    "enabled": Object {
      "message": "\\"enabled\\" is required",
      "type": "any.required",
    },
    "password": Object {
      "message": "\\"password\\" with value \\"___\\" fails to match the required pattern: /^[a-zA-Z0-9]{3,30}$/",
      "type": "string.pattern.base",
    },
    "tags": Object {
      "message": "\\"tags\\" is required",
      "type": "any.required",
    },
    "username": Object {
      "message": "\\"username\\" is required",
      "type": "any.required",
    },
  },
  "values": Object {},
}
`;

exports[`joiResolver should return all the errors from joiResolver when validation fails with \`validateAllFieldCriteria\` set to true 1`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "\\"birthYear\\" must be a number",
      "type": "number.base",
      "types": Object {
        "number.base": "\\"birthYear\\" must be a number",
      },
    },
    "email": Object {
      "message": "\\"email\\" is not allowed to be empty",
      "type": "string.empty",
      "types": Object {
        "string.empty": "\\"email\\" is not allowed to be empty",
      },
    },
    "enabled": Object {
      "message": "\\"enabled\\" is required",
      "type": "any.required",
      "types": Object {
        "any.required": "\\"enabled\\" is required",
      },
    },
    "password": Object {
      "message": "\\"password\\" with value \\"___\\" fails to match the required pattern: /^[a-zA-Z0-9]{3,30}$/",
      "type": "string.pattern.base",
      "types": Object {
        "string.pattern.base": "\\"password\\" with value \\"___\\" fails to match the required pattern: /^[a-zA-Z0-9]{3,30}$/",
      },
    },
    "tags": Object {
      "message": "\\"tags\\" is required",
      "type": "any.required",
      "types": Object {
        "any.required": "\\"tags\\" is required",
      },
    },
    "username": Object {
      "message": "\\"username\\" is required",
      "type": "any.required",
      "types": Object {
        "any.required": "\\"username\\" is required",
      },
    },
  },
  "values": Object {},
}
`;

exports[`joiResolver should return all the errors from joiResolver when validation fails with \`validateAllFieldCriteria\` set to true 2`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "\\"birthYear\\" must be a number",
      "type": "number.base",
      "types": Object {
        "number.base": "\\"birthYear\\" must be a number",
      },
    },
    "email": Object {
      "message": "\\"email\\" is not allowed to be empty",
      "type": "string.empty",
      "types": Object {
        "string.empty": "\\"email\\" is not allowed to be empty",
      },
    },
    "enabled": Object {
      "message": "\\"enabled\\" is required",
      "type": "any.required",
      "types": Object {
        "any.required": "\\"enabled\\" is required",
      },
    },
    "password": Object {
      "message": "\\"password\\" with value \\"___\\" fails to match the required pattern: /^[a-zA-Z0-9]{3,30}$/",
      "type": "string.pattern.base",
      "types": Object {
        "string.pattern.base": "\\"password\\" with value \\"___\\" fails to match the required pattern: /^[a-zA-Z0-9]{3,30}$/",
      },
    },
    "tags": Object {
      "message": "\\"tags\\" is required",
      "type": "any.required",
      "types": Object {
        "any.required": "\\"tags\\" is required",
      },
    },
    "username": Object {
      "message": "\\"username\\" is required",
      "type": "any.required",
      "types": Object {
        "any.required": "\\"username\\" is required",
      },
    },
  },
  "values": Object {},
}
`;
